includes/AMP/Output_Buffer.php

```diff
@@ -0,0 +1,295 @@
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories\AMP;
+
+use DOMElement;
+use Exception;
+use Google\Web_Stories\Experiments;
+use Google\Web_Stories\Model\Story;
+use Google\Web_Stories\Service_Base;
+use Google\Web_Stories\Settings;
+use Google\Web_Stories\Story_Post_Type;
+use Google\Web_Stories\Traits\Publisher;
+use Google\Web_Stories_Dependencies\AmpProject\Dom\Document;
+use Throwable;
+use WP_Post;
+
+/**
+ * Output buffer class.
+ *
+ * Largely copied from AMP_Theme_Support.
+ *
+ * @since 1.10.0
+ *
+ * @see \AMP_Theme_Support
+ */
+class Output_Buffer extends Service_Base {
+	use Publisher;
+
+	/**
+	 * Whether output buffering has started.
+	 *
+	 * @var bool
+	 */
+	protected $is_output_buffering = false;
+
+	/**
+	 * Experiments instance.
+	 *
+	 * @var Experiments Experiments instance.
+	 */
+	private $experiments;
+
+	/**
+	 * HTML constructor.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param Experiments $experiments Experiments instance.
+	 */
+	public function __construct( Experiments $experiments ) {
+		$this->experiments = $experiments;
+	}
+
+	/**
+	 * Runs on instantiation.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @return void
+	 */
+	public function register() {
+		// If the AMP plugin is installed and available in a version >= than ours,
+		// all sanitization and optimization should be delegated to the AMP plugin.
+		if ( defined( '\AMP__VERSION' ) && version_compare( AMP__VERSION, WEBSTORIES_AMP_VERSION, '>=' ) ) {
+			return;
+		}
+
+		/*
+		 * Start output buffering at very low priority for sake of plugins and themes that use template_redirect
+		 * instead of template_include.
+		 */
+		$priority = defined( 'PHP_INT_MIN' ) ? PHP_INT_MIN : ~PHP_INT_MAX; // phpcs:ignore PHPCompatibility.Constants.NewConstants.php_int_minFound
+		add_action( 'template_redirect', [ $this, 'start_output_buffering' ], $priority );
+
+		add_filter( 'web_stories_amp_sanitizers', [ $this, 'add_web_stories_amp_content_sanitizers' ] );
+	}
+
+	/**
+	 * Start output buffering.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @see Sanitization::finish_output_buffering()
+	 *
+	 * @return void
+	 */
+	public function start_output_buffering() {
+		if ( ! is_singular( Story_Post_Type::POST_TYPE_SLUG ) ) {
+			return;
+		}
+
+		/*
+		 * Disable the New Relic Browser agent on AMP responses.
+		 * This prevents the New Relic from causing invalid AMP responses due the NREUM script it injects after the meta charset:
+		 * https://docs.newrelic.com/docs/browser/new-relic-browser/troubleshooting/google-amp-validator-fails-due-3rd-party-script
+		 * Sites with New Relic will need to specially configure New Relic for AMP:
+		 * https://docs.newrelic.com/docs/browser/new-relic-browser/installation/monitor-amp-pages-new-relic-browser
+		 */
+		if ( function_exists( 'newrelic_disable_autorum' ) ) {
+			newrelic_disable_autorum();
+		}
+
+		ob_start( [ $this, 'finish_output_buffering' ] );
+		$this->is_output_buffering = true;
```

Comment:
We are doing this output buffering a couple of places now. Wonder if this should be a base class or similar to do this work....

Reply:
I don't think we need a base class for doing simple `ob_start()` calls in our codebase. Also, this class here is specific to AMP sanitization and the only place where we need a callback and an `is_output_buffering` flag.

Adding a level of abstraction on top of this seems overkill to me.